### PR TITLE
Refactor Crud

### PR DESCRIPTION
**Pull request by krisdo**
### What

- Refactored CRUD routes and CRUD tests

### Why

- Switched from MongoDB to PostgreSQL
- Readability, Modularity, Clean Code

### How & Where

- Wrote Query functions using knex (query builder) to insert, update, read, and delete to and from the database 
    1. `price_and_promotion/database/database-postgres/query-functions/insertRecords.js`
    2. `price_and_promotion/database/database-postgres/query-functions/updateRecords.js`
    3. `price_and_promotion/database/database-postgres/query-functions/getPriceandPromotions.js`
    4. `price_and_promotion/database/database-postgres/query-functions/deleteProductandDiscounts.js`
- The server routes takes parameters and run these functions
    1. `price_and_promotion/server/index.js`
- Tests refactored and all passed
    1. `price_and_promotion/tests/crud.test.js`
Screen Shot 2020-09-13 at 8 16 19 PM
UI is not broken! See Screenshot.
Screen Shot 2020-09-13 at 8 18 01 PM

### Notes
- Old code are in files marked with "not-in-use"
-TODOs: If time permits, would refactor server to have routes, controllers, and models

### CODE REVIEW

File: database/database-postgres/query-functions/getPriceandPromotions.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        )
+        .leftJoin(
+          db({gend: 'general_discounts'})
+          .select('gend.product_id', 'gend.discount')
+          .whereRaw('(? >= gend.start) AND (? <= gend.end)', [date, date])
+          .as('gend'), 'gend.product_id', 'p.id'
+        )
+      ]).as('pgd')
+    , 'pgd.id', 'subd.id')
+    .groupBy('subd.id', 'subd.base_price', 'subd.discount');
```
Comment (cywang117) [inline database/database-postgres/query-functions/getPriceandPromotions.js:53]: Complicated query! Yikes. Looks good, but I'm glad I don't have to stress test this.

File: server/index.js [hunk 2]
```diff
@@ -4,116 +4,147 @@ const app = express();
 const PORT = process.env.PORT || 3006;
 const path = require('path');
 const bodyParser = require('body-parser');
```
Comment (cywang117) [inline server/index.js:6]: If you're using Express v4, it has a built body-parser functionality via `app.use(express.json())`

File: server/index.js [hunk 3]
```diff
@@ -4,116 +4,147 @@ const app = express();
...
+// const { ESRCH } = require('constants');
+
+const getPriceAndPromotion = require('../database/database-postgres/query-functions/getPriceandPromotions.js');
+const deleteProductandDiscounts = require('../database/database-postgres/query-functions/deleteProductandDiscounts.js');
+const updateRecords = require('../database/database-postgres/query-functions/updateRecords.js');
+const insertRecords = require('../database/database-postgres/query-functions/insertRecords.js');
+const { EDESTADDRREQ } = require('constants');
 
 app.use(express.static('public'));
 app.use(bodyParser.urlencoded({ extended: true }));
```
Comment (cywang117) [inline server/index.js:19]: So you can get rid of these bodyParser lines if you want to :)

File: server/index.js [hunk 4]
```diff
@@ -4,116 +4,147 @@ const app = express();
...
+app.get('/PriceAndPromotion/multiple/:product_ids', async (req, res) => {
+
+  try {
+    if(!req.params.product_ids){
+      res.status(400).send('No Product Id');
+    } else if (req.params.product_ids) {
+      
+      let arrayOfIds = JSON.parse(req.params.product_ids);
+          
+      let arrayOfRecords = await getPriceAndPromotion(arrayOfIds);
```
Comment (cywang117) [inline server/index.js:67]: Nice flexibility in your get query for multiple/single ids

File: server/index.js [hunk 5]
```diff
@@ -4,116 +4,147 @@ const app = express();
...
+
+      let arrayOfPriceandPromtions = await arrayOfRecords.map( record => {
+
+        let {id, base_price, discount, max} = record;
+        let product_id = id;
+        let price = base_price;
+        let promotion = base_price * (discount/100);
+        promotion = Number(promotion.toFixed(2));
+        
+        if(max) {
```
Comment (cywang117) [inline server/index.js:77]: I noticed some inconsistent spacing that a linter would fix (lines 61, 63, here)

File: server/index.js [hunk 6]
```diff
@@ -4,116 +4,147 @@ const app = express();
...
-    .then( () => {
-      res.status(201).send('Post Successful');
-    })
-    .catch( (err) => {
-      res.status(404).send(err);
-    });
-  }
+  insertRecords(data)
+    .then( (response) => res.status(201).send({"NumberOfInsertedRecords": response.length, "ProductIds": response}))
+    .catch( (err) => res.status(400).send({error: JSON.stringify(err)}));
```
Comment (cywang117) [inline server/index.js:112]: Technically, a catch would take care of error handling, whether it's an incorrect post body or a db error, but maybe differentiating between the two with more granular error handling (code 400's vs 500's) would result in a better client experience / clarity? Just a thought!

File: server/index.js [hunk 7]
```diff
@@ -4,116 +4,147 @@ const app = express();
...
-//edits a record based on product id
-app.put('/PriceAndPromotion/:product_id', (req, res) => {
+//updates a record based on product id
+app.put('/PriceAndPromotion', (req, res) => {
 
-  let id = req.params.product_id;
-  let newData = req.body;
+  let data = req.body;
+
+  updateRecords(data)
```
Comment (cywang117) [inline server/index.js:120]: Same thought here as the post route comment.

File: tests/crud.test.js [hunk 8]
```diff
@@ -36,62 +37,105 @@ describe("Crud Operations", function(){
   })
```
Comment (cywang117) [inline tests/crud.test.js:38]: CRUD tests look good!

**Review by cywang117 — APPROVED** [review]
Looks good! The refactor to have your queries be in their own files is an instance of good modularity. Recommend setting up CI (very quick and painless, I promise) to avoid those pesky screenshots 👍

File: server/index.js (hunk 2 above)
Comment (krisdo) [reply server/index.js:6]: Thanks! This is inherited code so I didn't bother to change it. :X

**Pull request merged**